src/main/java/net/neoforged/neoforge/transfer/item/ItemResource.java

```diff
@@ -0,0 +1,126 @@
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.level.ItemLike;
+import net.neoforged.neoforge.transfer.Resource;
+
+/**
+ * Immutable combination of an {@link Item} and data components.
+ * Similar to an {@link ItemStack}, but immutable and without amount information.
+ */
+// TODO: probably a "hot take" but I actually prefer ItemVariant :P
+public final class ItemResource implements Resource, DataComponentHolder {
+    /**
+     * Codec for an item resource.
+     * Same format as {@link ItemStack#SINGLE_ITEM_CODEC}.
+     * Does <b>not</b> accept blank resources.
+     */
+    public static final Codec<ItemResource> CODEC = ItemStack.SINGLE_ITEM_CODEC
+            .xmap(ItemResource::of, ItemResource::toStack);
+    /**
+     * Codec for an item resource. Same format as {@link #CODEC}, and also accepts blank resources.
+     */
+    public static final Codec<ItemResource> OPTIONAL_CODEC = ExtraCodecs.optionalEmptyMap(CODEC)
+            .xmap(o -> o.orElse(ItemResource.EMPTY), r -> r.isBlank() ? Optional.of(ItemResource.EMPTY) : Optional.of(r));
+    // TODO
+//    /**
+//     * Codec for an item resource and an amount. Does <b>not</b> accept empty stacks.
+//     */
+//    public static final Codec<ResourceAmount<ItemResource>> WITH_AMOUNT_CODEC = ItemStack.CODEC
+//            .xmap(ItemStack::immutable, ItemStack::of);
+//    /**
+//     * Codec for an item resource and an amount. Accepts empty stacks.
+//     */
+//    public static final Codec<ResourceAmount<ItemResource>> OPTIONAL_WITH_AMOUNT_CODEC = ItemStack.OPTIONAL_CODEC
+//            .xmap(ItemStack::immutable, ItemStack::of);
+    /**
+     * Stream codec for an item resource. Accepts blank resources.
+     */
+    public static final StreamCodec<RegistryFriendlyByteBuf, ItemResource> OPTIONAL_STREAM_CODEC = StreamCodec.composite(
+            ByteBufCodecs.holderRegistry(Registries.ITEM),
+            ItemResource::getItemHolder,
+            DataComponentPatch.STREAM_CODEC,
+            ItemResource::getComponentsPatch,
+            ItemResource::of);
+
+    public static final ItemResource EMPTY = new ItemResource(ItemStack.EMPTY);
+
+    public static ItemResource of(ItemStack itemStack) {
+        return itemStack.isEmpty() ? EMPTY : new ItemResource(itemStack.copyWithCount(1));
+    }
+
+    public static ItemResource of(ItemLike item) {
+        return item == Items.AIR ? EMPTY : new ItemResource(new ItemStack(item));
+    }
+
+    public static ItemResource of(Holder<Item> item, DataComponentPatch patch) {
+        return item.value() == Items.AIR ? EMPTY : new ItemResource(new ItemStack(item, 1, patch));
+    }
+
+    /**
+     * We wrap an item stack which must never be exposed and/or modified.
+     */
+    private final ItemStack innerStack;
+
+    private ItemResource(ItemStack innerStack) {
+        this.innerStack = innerStack;
+    }
+
+    @Override
+    public boolean isBlank() {
```

Comment:
Can we get more utilities from `ItemStack` copied over?
mainly `ItemStack#isItemEnabled` and all of the `ItemStack#is` methods

This would allow consumers to run validation against a given `ItemResource` without needing to invoke `toStack` which creates a new copy of the backing `ItemStack`

Reply:
Yes, there is definitely room for more helpers and I suspect we'll be adding more over time.